### ISSUE TITLE
PyTorch/XLA GPU support

### ISSUE THREAD

**Issue by singam-sanjay** [pytorch/xla#694]
I would like to use PyTorch/XLA on GPUs.

But, PyTorch/XLA doesn't seem to work on them since XLA_GPU_DEVICE isn't registered, although xla_gpu_device.cpp seems to be included in the appropriate BUILD file since use_cuda==True in .bazelrc.

Thank You !

**Comment by dlibenzi** [comment]
Sorry, we do not support using GPU via XLA at this time.
The good news is that PyTorch GPU support is pretty good ;)

**Comment by singam-sanjay** [comment]
Thanks for responding ! Is GPU support part of your roadmap ?

**Comment by dlibenzi** [comment]
It is, especially because it should be a matter a very few code changes and build flags.

**Comment by singam-sanjay** [comment]
Thank You !

**Comment by wangzhimingchn** [comment]
@dlibenzi  Any update now? Can PyTorch/XLA GPU support use now?

**Comment by ailzhang** [comment]
Hi @wangzhimingchn, thanks for reaching out! So far PyTorch/XLA GPU is not in the roadmap. I'm curious what's your use case for it that normal Pytorch GPU backend cannot do? Thanks!

**Comment by wangzhimingchn** [comment]
@ailzhang Because of the performance of normal PyTorch GPU backend is not good. We transfered one of the models from PyTorch to TensorFlow, and run it with XLA, the performance can be updated. But we can't transfer all PyTorch models, so we need PyTorch/XLA GPU.
Any suggestion?

**Comment by dlibenzi** [comment]
Is the interest in running on a locally attached GPU, or running it on Google Cloud GPU?

**Comment by wangzhimingchn** [comment]
> on a locally attached GPU

On a locally attached GPU; I think it is because XLA can do more optimization when compiling the Graph.

**Comment by dlibenzi** [comment]
So, the good news is that we have got pytorch/xla built for using XLA_GPU, and basic tests passes.
It can even talk to different GPUs attached to the same host, but current XLA AllReduce on GPU (which use NVIDIA NCCL) has a rendezvous which only works for multiple threads within the same process, and only works with an XLA execution API which does not fit the pytorch/xla model:

https://github.com/tensorflow/tensorflow/blob/45d7117f0a9333aa676cc2af7c7e5cd32c7dbe3e/tensorflow/compiler/xla/service/gpu/nccl_all_reduce_thunk.cc#L43

This means that, even though pytorch/xla can see multiple GPU devices attached, the CroosReplicaSum based reduction we issue in pytorch/xla will not work.

**Comment by wangzhimingchn** [comment]
> So, the good news is that we have got pytorch/xla built for using XLA_GPU, and basic tests passes.
> It can even talk to different GPUs attached to the same host, but current XLA AllReduce on GPU (which use NVIDIA NCCL) has a rendezvous which only works for multiple threads within the same process, and only works with an XLA execution API which does not fit the pytorch/xla model:
> 
> https://github.com/tensorflow/tensorflow/blob/45d7117f0a9333aa676cc2af7c7e5cd32c7dbe3e/tensorflow/compiler/xla/service/gpu/nccl_all_reduce_thunk.cc#L43
> 
> This means that, even though pytorch/xla can see multiple GPU devices attached, the CroosReplicaSum based reduction we issue in pytorch/xla will not work.

Great! I really look forward to it.
This meas I can use PyTorch/XLA for training or inferencing on Single GPU now, is it?

**Comment by dlibenzi** [comment]
Yes, for now, even though pytorch/xla can see and use all the GPU devices attached to an host, XLA gradient reduction will not work.

**Comment by wangzhimingchn** [comment]
@dlibenzi I see. Thanks a lot!
